src/workergroup.ts

```diff
@@ -106,34 +111,39 @@ export class WorkerGroup {
 
   public setupSolver(forceJS = false) {
     const msg: SolverMessage = { type: "solver", forceJS: forceJS };
-    for (let i=0; i<this.workers.length; i++) {
+    for (let i = 0; i < this.workers.length; i++) {
       this.workers[i].postMessage(msg);
     }
   }
 
-  public start(puzzle: Puzzle) {
-    this.puzzleSolverInputs = getPuzzleSolverInputs(puzzle.buffer, puzzle.n);
+  private async getReadyWorker(): Promise<Worker> {
+    while (!this.readyWorkers.length) {
+      await new Promise<void>((resolve) => (this.resolveNextReady = resolve));
```

Comment:
I'm not sure what is happening here, is this equivalent an 'await' chain? By that I mean something like:
```javascript
this.prom = Promise.resolve()
// ...
this.prom = this.prom.then(async () => {..})
```
With something like that you end up with the work happening exactly in order and serially (which I guess is not what we want here). I think what is happening here is that we are waiting for any worker to be available, so they woud resolve this promise?

---
Have you considered perhaps having the following approach?
* We keep a `todo` variable.
* We start N workers, who we feed one of the todo's.
* Whenever a worker sends us their "done" message (in a closure, so we know where to send it back), we feed them another todo. If todo is empty we don't give another job.

A bit less state to handle with that setup and no need for 'inverted' promises. It's possible I'm not seeing some requirement that makes this approach not viable though! Interested to hear your thoughts.